### PR TITLE
Fix #414: Crash when re-applying time signature to parts (out of bounds)

### LINKED ISSUE #414: Remove first bar causes a crash
If I:
- select the first bar
-  right click
- remove the bar
Musescore 3.7 crashes. No problems with MS 3.6.2 or 4.X
The crash occurs if the first bar is empty

### PR DESCRIPTION
**Pull request by worldwideweary**
Resolves: #414

Test it out, seems to work okay

### CODE REVIEW

File: libmscore/edit.cpp [hunk 1]
```diff
@@ -2252,12 +2252,15 @@ void Score::deleteMeasures(MeasureBase* is, MeasureBase* ie, bool preserveTies)
...
-                        TimeSig* nts = new TimeSig(this);
-                        nts->setTrack(staffIdx * VOICES);
-                        nts->setParent(s);
-                        nts->setFrom(lastDeletedForThisStaff);
-                        nts->setStretch(nts->sig() / mAfterSel->timesig());
-                        score->undoAddElement(nts);
+
+                        auto track = staffIdx * VOICES;
+                        if (track < s->elist().size()) {
+                              TimeSig* nts = new TimeSig(score);
```
Comment (worldwideweary) [inline libmscore/edit.cpp:2258]: Two changes here:
1) verifying track is okay in relation to the score (in a loop going through all sub score or parts) (This gets verified in checkElement() function but it was asserting and crashing ungracefully
2) the new TimeSig is now getting the score for the part and not the this pointer which was the **master score**

**Comment by worldwideweary** [comment]
Obviously I'm not spot on because there's

> 27 - tst_measure (Failed)

... don't have local testing here though so not sure what that is yet

File: libmscore/edit.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [reply libmscore/edit.cpp:2258]: Seems the same change is needed for Mu4?

**Comment by Jojo-Schmitz** [comment]
What's the deal with the failing mtest?

File: libmscore/edit.cpp (hunk 1 above)
Comment (worldwideweary) [reply libmscore/edit.cpp:2258]: Not sure, it's possible Mu4 has an appropriate check mechanism somewhere else down the line that would render this redundant...

**Comment by worldwideweary** [comment]
> What's the deal with the failing mtest?

I dunno man. I don't know much about the unit testing... so I'm curious myself.

**Comment by worldwideweary** [comment]
There's the defined constructor which this code calls, and the old one which used the "default" constructor which doesn't have any code here... the one this uses and the only one showing in the codebase does:

```cpp
TimeSig::TimeSig(Score* s)
  : Element(s, ElementFlag::ON_STAFF | ElementFlag::MOVABLE)
      {
      initElementStyle(&timesigStyle);

      _showCourtesySig = true;
      _stretch.set(1, 1);
      _sig.set(0, 1);               // initialize to invalid
      _timeSigType      = TimeSigType::NORMAL;
      _largeParentheses = false;
```
So maybe these default attributes are not in accord with the default construction in some way that alters the mtest results...

**Comment by Jojo-Schmitz** [comment]
>       _sig.set(0, 1);               // initialize to invalid

Esp. this
```
--- /home/runner/work/MuseScore/MuseScore/mtest/libmscore/measure/measure-6-ref.mscx	2024-09-12 10:58:03.497486946 +0000
+++ measure-6.mscx	2024-09-12 11:19:15.121886888 +0000
@@ -59,8 +59,10 @@
       <Measure>
         <voice>
           <TimeSig>
-            <sigN>4</sigN>
-            <sigD>4</sigD>
+            <sigN>0</sigN>
+            <sigD>1</sigD>
+            <stretchN>0</stretchN>
+            <stretchD>1</stretchD>
             </TimeSig>
           <Chord>
             <durationType>quarter</durationType>
```

**Comment by Jojo-Schmitz** [comment]
Here's what that mtest does:
```cpp
void TestMeasure::spanner_A()
      {
      MasterScore* score = readScore(DIR + "measure-6.mscx");

      score->select(score->firstMeasure());
      score->startCmd();
      score->localTimeDelete();
      score->endCmd();
      score->doLayout();
      QVERIFY(saveCompareScore(score, "measure-6.mscx", DIR + "measure-6-ref.mscx"));
      delete score;
      }
```

**Comment by worldwideweary** [comment]
You know what, I think I see what the problem is already... The results of the dude's score gives: 

image

0/1 

Something's wrong elsewhere...

**Comment by worldwideweary** [comment]
Oh, it's the stretch:

nts->setStretch(nts->sig() / mAfterSel->timesig());

wtf is that doing?

File: libmscore/edit.cpp [hunk 2]
```diff
@@ -2252,12 +2252,15 @@ void Score::deleteMeasures(MeasureBase* is, MeasureBase* ie, bool preserveTies)
...
                               s = mAfterSel->undoGetSegment(SegmentType::TimeSig, mAfterSel->tick());
 
-                        TimeSig* nts = new TimeSig(this);
-                        nts->setTrack(staffIdx * VOICES);
-                        nts->setParent(s);
-                        nts->setFrom(lastDeletedForThisStaff);
-                        nts->setStretch(nts->sig() / mAfterSel->timesig());
-                        score->undoAddElement(nts);
+                        auto track = staffIdx * VOICES;
+                        if (track < static_cast<int>(s->elist().size())) {
```
Comment (Jojo-Schmitz) [inline libmscore/edit.cpp:2256]: maybe better change that `auto` to `size_t`

Comment (worldwideweary) [reply libmscore/edit.cpp:2256]: probably right. no casting needed then

File: libmscore/edit.cpp (hunk 1 above)
Comment (Jojo-Schmitz) [reply libmscore/edit.cpp:2257]: Now a 3rd change `setFrom(lastDeletedForThisStaff);` to `setSig(lastDeletedForThisStaff->sig());`

**Comment by worldwideweary** [comment]
OK, hopefully that takes care of it.
 If there are any problems in the future, alternatively the code could just the "this" pointer like originally and then specifically set set the score afterwards.

**Pull request merged**